DasharoPayloadPkg/CapsuleSplashDxe/CapsuleSplashDxe.c

```diff
@@ -0,0 +1,345 @@
+  //
+  mFBuf.x += left;
+  mFBuf.y += top;
+
+  //
+  // Render the text. The function takes care to not write outside of buffer.
+  // It returns number of bytes consumed to render one glyph (UTF-8). Special
+  // characters like \n are not handled.
+  //
+  while((Status = ssfn_render(&mCtx, &mFBuf, str)) > 0)
+    str += Status;
+
+  if (Status < 0) {
+    DEBUG ((DEBUG_ERROR, "ssfn_render failed with %a!\n", ssfn_error(Status)));
+    return EFI_UNSUPPORTED;
+  }
+
+  return EFI_SUCCESS;
+}
+
+/**
+  The Entry Point for CapsuleSplash driver.
+
+  @param[in] ImageHandle    The firmware allocated handle for the EFI image.
+  @param[in] SystemTable    A pointer to the EFI System Table.
+
+  @retval EFI_SUCCESS       The entry point is executed successfully.
+  @retval other             Some error occurs when executing this entry point.
+
+**/
+EFI_STATUS
+EFIAPI
+CapsuleSplashEntry (
+  IN EFI_HANDLE                        ImageHandle,
+  IN EFI_SYSTEM_TABLE                  *SystemTable
+  )
+{
+  EFI_STATUS                            Status;
+  EFI_GRAPHICS_OUTPUT_PROTOCOL          *GraphicsOutput;
+  EFI_GRAPHICS_OUTPUT_MODE_INFORMATION  *ModeInfo;
+
+  //
+  // Find GOP and ensure that pixel size is 32b.
+  //
+  Status = gBS->HandleProtocol (gST->ConsoleOutHandle,
+                                &gEfiGraphicsOutputProtocolGuid,
+                                (VOID **)&GraphicsOutput);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "Couldn't find GOP\n"));
+    return Status;
+  }
+
+  ModeInfo = GraphicsOutput->Mode->Info;
+  if (ModeInfo->PixelFormat != PixelRedGreenBlueReserved8BitPerColor &&
+      ModeInfo->PixelFormat != PixelBlueGreenRedReserved8BitPerColor) {
+    DEBUG ((DEBUG_ERROR, "Wrong pixel format\n"));
+    return EFI_UNSUPPORTED;
+  }
+
+  //
+  // Set dummy logo for consistent progress bar
+  //
+  Status = SetDummyLogo (ModeInfo);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "Couldn't set dummy logo\n"));
+    return Status;
+  }
+
+  //
+  // Initialize font renderer.
+  //
+  Status = InitSsfn (GraphicsOutput->Mode->FrameBufferBase, ModeInfo);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "Couldn't init font renderer\n"));
+    return Status;
+  }
+
+  //
+  // Clear the background. We don't know what logo was shown there, if any.
+  //
+  SetMem ((VOID *)GraphicsOutput->Mode->FrameBufferBase,
+          GraphicsOutput->Mode->FrameBufferSize,
+          0x00);
+
+  //
+  // Print some warnings.
+  //
+  Status = RenderTextCenteredAt("Firmware update in progress",
+                                ModeInfo->HorizontalResolution / 2,
+                                ModeInfo->VerticalResolution * 20 / 100);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  Status = RenderTextCenteredAt("Don't turn off your platform!",
+                                ModeInfo->HorizontalResolution / 2,
+                                ModeInfo->VerticalResolution * 90 / 100);

```

Comment:
I think these percentages could be adjusted (this one to 88 and above one to 37) to align things better and instead of

![](https://private-user-images.githubusercontent.com/40995177/370257685-67a1be33-841d-4db0-8b5c-db390911a3f2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcyMTMzMjcsIm5iZiI6MTcyNzIxMzAyNywicGF0aCI6Ii80MDk5NTE3Ny8zNzAyNTc2ODUtNjdhMWJlMzMtODQxZC00ZGIwLThiNWMtZGIzOTA5MTFhM2YyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI0VDIxMjM0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMxMDcyOTBjODE0NmRjYzEzZTQ0ZmM2ODczZDdmNjVhN2MzZGMzMjFlZjlhMmQ5MTcwODQ0MTdiOTdkZTQ3Y2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.9QLfc7peQR-tSAsgb8hmu6hOJkcDyNlUZlU2kGa4NTE) get 
![ux](https://github.com/user-attachments/assets/c70c868c-e7db-4e51-863a-e9768259127f)

Reply:
Moved in https://github.com/Dasharo/edk2/pull/167/commits/1382f900afd5dc6ee950833400f15175f6840654#diff-8abe742e2056f5e030b724bbaf8b66e9b48ab6494be416fd65f3a10cd743b371R336 and https://github.com/Dasharo/edk2/pull/167/commits/1382f900afd5dc6ee950833400f15175f6840654#diff-8abe742e2056f5e030b724bbaf8b66e9b48ab6494be416fd65f3a10cd743b371R340.

I planned to add a glyph for Dasharo logo and print it in the middle, but it requires simplified SVG (just lines and Bezier curves, no arcs etc.) While working on it I discovered that printing the text would return the error on too big font size, and while it would require enormous resolution to trigger for existing lines, I added the code to clamp it anyway: https://github.com/Dasharo/edk2/pull/167/commits/1382f900afd5dc6ee950833400f15175f6840654#diff-8abe742e2056f5e030b724bbaf8b66e9b48ab6494be416fd65f3a10cd743b371R169-R172

Reply:
Out of curiosity, why did you try to draw it as a glyph instead of a bitmap as the logo is drawn?

Reply:
Just for proper scaling. We will soon have a platform with 2880x1800 display, the usual logo would be too small for that.